apps/prairielearn/src/pages/instructorAssessmentAccessOverrides/instructorAssessmentAccessOverrides.js

```diff
@@ -0,0 +1,149 @@
+        });
+        // Get the group_id from the result
+        if (group_result.rows.length > 0) {
+          params.group_id = group_result.rows[0].id;
+        } else {
+          params.group_id = null;
+        }
+
+        // If group does not belong to assessments and indirectly course instances, return error
+        if (!params.group_id) {
+          throw error.make(400, 'Group not found in this assessment.');
+        }
+      } else {
+        if (!params.user_id || params.group_name) {
+          throw error.make(400, 'Student User ID is required for individual work assessments.');
+        }
+      }
+      await sqldb.queryAsync(sql.insert_assessment_access_policy, params);
+      res.redirect(req.originalUrl);
+    } else if (req.body.__action === 'delete_override') {
+      await sqldb.queryAsync(sql.delete_assessment_access_policy, {
+        assessment_id: res.locals.assessment.id,
+        unsafe_assessment_access_policies_id: req.body.policy_id,
+      });
+      res.redirect(req.originalUrl);
+    } else if (req.body.__action === 'edit_override') {
+      let user_id = null;
+      if (req.body.student_uid) {
+        user_id = await getUserIdAndCheckEnrollment({
+          uid: req.body.student_uid,
+          course_instance_id: res.locals.course_instance.id,
+        });
+      }
+
+      const edit_params = {
+        assessment_id: res.locals.assessment.id,
+        credit: req.body.credit,
+        end_date: new Date(req.body.end_date),
+        group_name: req.body.group_name || null,
+        note: req.body.note || null,
+        start_date: new Date(req.body.start_date),
+        group_id: null,
+        user_id: user_id || null,
+        assessment_access_policies_id: req.body.policy_id,
+      };
+
+      // Validate if group belongs to the assessment, otherwise check if student is enrolled in assessment
+      if (res.locals.assessment.group_work) {
+        if (!edit_params.group_name || edit_params.user_id) {
+          throw error.make(400, 'Group name is required for group work assessments.');
+        }
+        const group_result = await sqldb.queryAsync(sql.select_group_in_assessment, {
+          group_name: edit_params.group_name,
+          course_instance_id: res.locals.course_instance.id,
+          assessment_id: res.locals.assessment.id,
+        });
+        // Get the group_id from the result
+        if (group_result.rows.length > 0) {
+          edit_params.group_id = group_result.rows[0].id;
+        } else {
+          edit_params.group_id = null;
+        }
+
+        // If group does not belong to assessments and indirectly course instances, return error
+        if (!edit_params.group_id) {
+          throw error.make(400, 'Group does not belong to the current course instance.');
+        }
+      } else {
+        if (!edit_params.user_id || edit_params.group_name) {
+          throw error.make(400, 'Student UID is required for individual work assessments.');
+        }
+      }
```

Comment:
Nit: from what I can tell, this code (and the `getUserIdAndCheckEnrollment` code above) are repeated pretty much verbatim in the "create" and "edit" cases. You might consider lifting these into a shared abstraction that could be used something like the following:

```ts
const { user_id, group_id } = await getUserOrGroupId(assessment, req.body.student_uid, req.body.group_name);
await sqldb.queryAsync(sql.update_assessment_access_policy, {
  assessment_id: res.locals.assessment.id,
  // ...
  user_id,
  group_id,
});
```

Reply:
The code is so much easier to read now, lifting into a shared abstraction did wonders!

Reply:
Yes, it's looking much better now!